Scripts/Playthrough.cs

```diff
@@ -374,6 +341,208 @@ private bool IsConnected()
+                }
+            });
+
+            _room.OnMessage<Events.MessageEvent>("message", (message) =>
+            {
+                Logger.Log($"Received `message` event: {JsonConvert.SerializeObject(message)}");
+                OnMessage?.Invoke(message);
+            });
+
+            _room.OnMessage<Events.StartTypingEvent>("start-typing", (message) =>
+            {
+                Logger.Log($"Received `start-typing` event: {JsonConvert.SerializeObject(message)}");
+                OnStartTyping?.Invoke(message);
+            });
+
+            _room.OnMessage<Events.StopTypingEvent>("stop-typing", (message) =>
+            {
+                Logger.Log($"Received `stop-typing` event: {JsonConvert.SerializeObject(message)}");
+                OnStopTyping?.Invoke(message);
+            });
+
+            _room.OnMessage<Events.ProblemEvent>("problem", (message) =>
+            {
+                Logger.LogWarning($"Received `problem` event: {JsonConvert.SerializeObject(message)}");
+            });
+
+            _room.OnMessage<string>("pong", (message) =>
+            {
+                _pingCount = 0;
+                OnPingSuccess?.Invoke();
+                SetConnectionState(ConnectionState.Connected);
+            });
+
+            _room.OnLeave += (code) =>
+            {
+                Logger.Log($"Connection closed. Attempting to reconnect to Playthrough.");
+                MainThreadConsumer.Instance?.StartCoroutine(TryToReconnect());
+            };
+        }
+
+        private void OnReady()
+        {
+            SetConnectionState(ConnectionState.Connected);
+
+            _onReadyCallback?.Invoke();
+
+            _calledByDisconnect = false;
+            _reconnectionTryCount = 0;
+
+            Task.Run(FirePing);
+        }
+
+        private IEnumerator TryToReconnect()
+        {
+            Logger.Log("Trying to reconnect.");
+            if (_calledByDisconnect)
+            {
+                SetConnectionState(ConnectionState.Disconnected);
+                yield break;
+            }
+
+            // cannot allow multiple reconnection coroutines to run concurrently
+            if (IsReconnecting())
+            {
+                yield break;
+            }
+
+            SetConnectionState(ConnectionState.Reconnecting);
+
+            MainThreadConsumer.Instance.StartCoroutine(Reconnect());
+        }
+
+        private bool IsReconnecting()
+        {
+            return _connectionState == ConnectionState.Reconnecting;
+        }
+
+        private IEnumerator Reconnect()
+        {
+            ColyseusRoom<object> room = null;
+
+            while (room == null)
+            {
+                // Try to reconnect
+                _reconnectionTryCount++;
+
+                if (_reconnectionTryCount <= MAXIMUM_RECONNECTION_ATTEMPTS)
+                {
+                    _client.Settings.useSecureProtocol = true;
+                    var t = _client.Reconnect(_room.Id, _room.SessionId);
```

Comment:
usually when reconnecting, we can get away with just using `Reconnect()` as it is here, which works as long as the playthrough 'room' still exists on the server

however sometimes, especially when a server updates, the playthrough room is dropped entirely and the client needs to request to re-create the room

we can test if that's the case because `_client.Reconnect()` will throw an error which looks something like `room "ROOM_ID" not found`

in this scenario we need to call `client.JoinOrCreate()` in the same way we do initially, it may be easier to just call `CreateClientAndJoinRoom` again

so in a nutshell:

- keep trying to `_client.Reconnect()`
- if `_client.Reconnect()` fails with an error matching the string `room "ROOM_ID" not found` then instead...
- keep trying to `_client.JoinOrCreate()`

(also rename `t` to something more descriptive like `reconnectionTask`!)

reference in JS SDK: https://github.com/charisma-ai/charisma-sdk-js/blob/ac3c243d5e79be654d91aa7d03363e3d19721e5a/src/Playthrough.ts#L270